design/034-new_lockfiles.md

```diff
@@ -0,0 +1,237 @@
+A ``conan lock add`` command will be provided, so it is more convenient to add things to lockfiles.
+
+Furthermore, it will also be possible to define partial references, that is, even if a lockfile by default captures the ``pkg/version@user/channel#recipe-revision`` full recipe reference, it is still possible to define with ``conan lock add`` a given ``version`` without specifying a recipe revision. In that case, when the lockfile is used, it will use the locked version, and it will resolve dynamically (to latest) the missing piece of recipe revision.
+
+
+### Modifying lockfiles
+
+Let's say we have a dependency graph ``pkgb``->``pkga``.
+
+We can capture first a lockfile for the dependencies of ``pkgb``, while we are developing it. It will contain only the locked version of ``pkga``. If we decide to ``conan create`` the ``pkgb`` using those locked dependencies, we can augment the existing lockfile to include the recently created ``pkgb`` version (and recipe revision) to the same or to a new lockfile:
+
+```bash
+# in the pkgb/conanfile, assuming it ``requires = “pkg/[>0.0]”``
+conan lock create . --lockfile-out=pkgb.lock
+# this pkgb.lock will contain a locked reference to say ``pkga/0.1``
+conan create . --lockfile=pkgb.lock --lockfile-out=pkgb.lock
+# now the pkgb.lock contains ALSO the locked pkgb version
+```
+
+We have commented above that a ``conan lock add`` operation will be possible, to manually add a version. In the same spirit, it will also be possible to ``conan lock merge`` different lockfiles, as long as those lockfiles belong to the same dependency graph, partially or totally. But they should have shared some common root at some point in time.
+
+These modifications basically add new versions to an existing lockfile, but as such lockfile can be used for different configurations, it doesn’t remove non-used locked references automatically. In order to clean and strip unused references from a lockfile, an explicit ``conan lock create --clean`` creation of a lockfile will be possible.
+
+### Lockfiles in CI
+
+This proposal simplifies the lockfiles and removes some of the commands and functionality that Conan 1.X lockfiles implemented, and that is necessary for implementing CI flows. Conan 2.0 will enable these CI flows with the following functionality:
+
+- ``conan graph build-order`` will output the necessary build-order for a full dependency graph, returning an ordered list of lists, by concurrent-builds levels, in which things need to be built.
+- ``conan graph build-order`` will also return the necessary options that every given package in the build-order will need to apply (this information was previously hidden in the lockfile)
+- It will be possible to merge different build-orders, even from different products and different dependency graphs with ``conan graph build-order-merge`` which will result in a grouped view, for every package reference, in order, which binaries need to be built.
```

Comment:
To merge 2 build orders 2 graphs are needed =>  2 recipes. Lockfiles2.0 don't contain graph, do they? And build order is an operation on graph(s) which are no longer in lockfiles.

Or do I miss something?

Reply:
Not necessarily 2 recipes, but maybe the same recipe with 2 configurations, you obtain 2 different dependency graphs, and they can result in different build orders, when conditional dependencies are defined for example. 

Yes, lockfiles no longer contains the graph definition, only lists of dependencies.

The key is that the build-order is no longer an operation on the lockfile itself (as it was in 1.X), but on the graph. Exactly the same as the ``conan graph info`` command (the current ``conan info`` command in 1.X), they build the entire dependency graph, without necessarily fetching the binaries. And the ``conan graph info`` and ``conan graph build-order`` can operate perfectly without a lockfile, in the same way a ``conan create`` or a ``conan install``. The ``build-order`` has been nicely decoupled from the lockfile definition, and the responsibilities are better defined now: the responsibility of a lockfile is to lock versions and revisions, when other Conan commands are executed.

Please let me know if this clarifies it.

Reply:
My main concern was how will UI / CLI for ``conan graph build-order-merge`` will look like. 
It is clear that for ``conan graph build-order`` one needs a complete graph (which is produced from recipe and profiles), I'm just not sure if it is possible to merge 2 lists of lists in a general case.

Reply:
The output of ``conan graph build-order`` is a list of lists, but it also contains dependencies information, each element in the list (as serialized to the json output file) is like:
```
{
                "ref": "pkg/0.1#1ac8dd17c0f9f420935abd3b6a8fa032",
                "depends": [
                    "dep/0.1#4d670581ccb765839f2239cc8dff8fbd"
                ],
                "packages": [
                    {
                        "package_id": "486166899301ccd88a8b71715c97eeea5cc3ff2b",
                        'prev': None,
                        'filenames': ["bo1"],
                        "context": "host",
                        "binary": "Build",
                        "options": []
                    }
                ]
            },
```

And then, multiple build-order files can be merged into a single one (typically resulting in more than one item in the ``packages`` section), as long as the ``depends`` are respected.

Reply:
Ah, that makes sense. So in a way you moved graph from lockfile into build-order output. Since the latter needs to be processed by a tool, I don't see many problems with that. Except maybe that previous build-order output could be parsed by simpler regexp, while for that one a complete json parser is needed. So bash users could be unhappy 😁 (or they will have to use tools like ``jq``).

Reply:
Yes, exactly. The UI was weird in 1.X, in the sense that you really shouldn't need a lockfile in order to compute the build-order of a graph. So it was moved to the right location, ``conan graph build-order``, which can work without lockfiles.

Yes, you are very right regarding the full json output, but from what we have learned from users so far is that such json is being parsed by CI in the vast majority of cases (because the list-of-lists is not that simple to parse with regex either). And in 2.0 for those who want something simpler, it will be possible very easily, for example with a user custom command, you can process the build-order and output exactly in the format that you want to process it, with very few lines of code, and integrated and managed by Conan.